xapian-core/cluster/kmeans.cc

```diff
@@ -0,0 +1,142 @@
+/** @file kmeans.cc
+ *  @brief KMeans clustering API
+ */
+/* Copyright (C) 2016 Richhiey Thomas
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/cluster.h"
+
+#include "debuglog.h"
+#include "xapian/error.h"
+#include "omassert.h"
+
+#include <climits>
+#include <random>
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+// Threshold value for checking convergence in KMeans
+#define CONVERGENCE_THRESHOLD 0.0000000001
+
+/** Maximum number of times KMeans algorithm will iterate
+ *  till it converges
+ */
+#define MAX_ITERS 1000
+
+using namespace Xapian;
+using namespace std;
+
+KMeans::KMeans(unsigned int k_, unsigned int max_iters_)
+    : k(k_) {
+    LOGCALL_CTOR(API, "KMeans", k_ | max_iters_);
+    max_iters = (max_iters_ == 0) ? MAX_ITERS : max_iters_;
+    if (k_ == 0)
+	throw InvalidArgumentError("Number of required clusters should be greater than zero");
+}
+
+string
+KMeans::get_description() const
+{
+    return "KMeans()";
+}
+
+void
+KMeans::initialize_clusters(ClusterSet &cset)
+{
+    LOGCALL_VOID(API, "KMeans::initialize_clusters", cset);
+    // Since we are using the Mersenne Twister 19937 pseudo-random number
+    // generator, the period will be greater than the value of k
+    // (period of mt19937 >> k). So we do not need to check whether
+    // the same point has been selected more than once
+    mt19937 generator;
+    generator.seed(random_device()());
+    uniform_int_distribution<mt19937::result_type> random(0, points.size() - 1);
+    for (unsigned int i = 0; i < k; ++i) {
+	int x = random(generator);
+	cset.add_cluster(Cluster(Centroid(points[x])));
+    }
```

Comment:
As I mentioned before, the current algorithm here is flawed, because the random number sequence can return the same random value in much less than the period of the generator.  To genuinely implement a random selection of `k` elements from `N` (where `N = points.size()`) seems like it's going to need `O(N)` temporary storage to ensure no duplicate selections.  At least the obvious approaches do - perhaps there's some cunning approach which doesn't.

Having thought about this, I wonder if it's simplest to just pick elements `i * points.size() / k` for i in {0, ..., k-1} as the initial centroids, instead of randomly choosing.  We're clustering an `MSet` which is sorted by either relevance or some sort key, so picking elements spread through the MSet should tend to pick diverse elements.